core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -629,15 +657,42 @@ public ShareAcquiredRecords acquire(
             if (subMap.isEmpty()) {
                 log.trace("No cached data exists for the share partition for requested fetch batch: {}-{}",
                     groupId, topicIdPartition);
-                return acquireNewBatchRecords(memberId, fetchPartitionData.records.batches(),
+                ShareAcquiredRecords acquiredRecords =  acquireNewBatchRecords(memberId, fetchPartitionData.records.batches(),
                     firstBatch.baseOffset(), lastBatch.lastOffset(), batchSize, maxFetchRecords);
+                // Since new records have been acquired, the window tracking the gap in cachedState might need to be reduced
+                maybeUpdateReadGapFetchOffset(lastBatch.lastOffset() + 1);
```

Comment:
This is incorrect. As the changes I made, you can see [here](https://github.com/apache/kafka/compare/trunk...apoorvmittal10:kafka:KAFKA-18494), that change goes inside the  `acquireNewBatchRecords` because if `maxFetchRecords` is configured then not neccessarily all fetch records will be acquired.

Reply:
Also you should write a test case for this as well.
cc: @adixitconfluent